File: src/cviz/control/tcp/ControlClient.java
package cviz.control.tcp;

import com.google.gson.Gson;
import com.google.gson.JsonIOException;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonWriter;
import cviz.TimelineManager;
import cviz.state.State;

import java.io.*;
import java.net.Socket;
import java.util.Arrays;

public class ControlClient implements Runnable {
    private final TimelineManager manager;
    private final Gson gson = new Gson();

    private Socket socket;
    private boolean connected;

    private final JsonReader inputStream;
    private final JsonWriter outputStream;
    private final OutputStream rawOutputStream;

    public ControlClient(TimelineManager manager, Socket socket) throws IOException {
        this.manager = manager;
        this.socket = socket;
        this.connected = true;

        try {
            rawOutputStream = socket.getOutputStream();
            outputStream =  new JsonWriter(new OutputStreamWriter(rawOutputStream, "UTF-8"));
            inputStream = new JsonReader(new InputStreamReader(socket.getInputStream(), "UTF-8"));
            inputStream.setLenient(true);
        } catch (IOException e) {
            System.err.println("Socket failed");
            try {
                socket.close();
            } catch (Exception e1) {
            }

            throw e;
        }
    }

    private void runAction(ClientAction action){
        switch(action.getType()){
            case KILL:
                manager.killTimeline(action.getName());
                break;

            case LOAD:
                String instanceId = action.getTemplateDataId() != null ? action.getTemplateDataId() : "";
                if(manager.loadTimeline(action.getChannel(), action.getName(), action.getFilename(), instanceId)) {
                    manager.startTimeline(action.getName(), action.getTemplateData());
                }
                break;

            case CUE:
                manager.triggerCue(action.getName());
                break;

            case QUERY:
                // Nothing to do
                break;

            default:
                System.err.println("Unknown action type: "+action.getType());
                break;
        }
    }

    @Override
    public void run() {
        while (isConnected()) {
            try {
                if(inputStream.hasNext()) {
                    ClientAction action = gson.fromJson(inputStream, ClientAction.class);
                    if(action == null || action.getType() == null) {
                        replyPing();
                    } else if (action.getType() == ClientAction.ActionType.QUERY) {
                        System.out.println("Received state query");
                        sendState(manager.getStateForTimelineId(action.getName()));
                    } else {
                        System.out.println("Received action: " + action);
                        runAction(action);
                    }
                } else {
                    try {
                        Thread.sleep(10);
                    } catch (InterruptedException e) {
                    }
                }

            } catch (Exception e) {
                System.err.println("Lost Connection to peer " + e.getMessage());
                // e.printStackTrace();
                close();
            }
        }
    }

    public void close() {
        synchronized (outputStream) {
            connected = false;
            try {
                if (socket != null)
                    socket.close();
                System.err.println("Closing Connection");
            } catch (IOException e) {
            }
            socket = null;
        }
    }

    boolean sendState(State state) {
        if (state == null)
            return true;

        if (!isConnected())
            return false;

        synchronized (outputStream) {
            try {
                gson.toJson(state, State.class, outputStream);
                outputStream.flush();
            } catch (JsonIOException je){
                System.err.println("Failed to send message: " + je.getMessage());
            } catch (Exception ioe) {
                System.err.println("Failed to send message: " + ioe.getMessage());
                if (socket != null) {
                    try {
                        socket.close();
                    } catch (IOException e) {
                    }
                }
                socket = null;
            }
        }

        return true;
    }

    private boolean replyPing() {
        if (!isConnected())
            return false;

        Object obj = new Object();

        new Thread(() -> {
            synchronized (outputStream) {
                try {
                    gson.toJson(obj, Object.class, outputStream);
                    outputStream.flush();
                } catch (JsonIOException je){
                    System.err.println("Failed to send ping: " + je.getMessage());
                } catch (Exception ioe) {
                    System.err.println("Failed to send ping: " + ioe.getMessage());
                    if (socket != null) {
                        try {
                            socket.close();
                        } catch (IOException e) {
                        }
                    }
                    socket = null;
                }
            }
        }).start();

        return true;
    }

    public boolean isConnected() {
        return socket != null && connected && socket.isConnected();
    }

    void sendCompleteState() {
        State[] state = manager.getCompleteState();

        Arrays.stream(state).forEach(this::sendState);
    }
}

File: CViz/Control/Tcp/ControlClient.cs
using System;
using System.Collections.Immutable;
using System.IO;
using System.Net.Sockets;
using System.Text;
using System.Threading.Tasks;
using CViz.Timeline;
using CViz.Util;
using log4net;
using Newtonsoft.Json;

namespace CViz.Control.Tcp
{
    class ControlClient
    {
        private static readonly ILog Log = LogManager.GetLogger(typeof(ControlClient));

        private readonly TimelineManager _manager;
        private Socket _socket;
        private readonly object _sendLock;
        private bool _connected;
        
        private const int BufferSize = 4096;
        private readonly byte[] _buffer = new byte[BufferSize];
        private readonly StringBuilder _sb = new StringBuilder();

        public ControlClient(TimelineManager manager, Socket socket)
        {
            _manager = manager;
            _socket = socket;
            _connected = true;

            _sendLock = new object();
        }

        public void Run()
        {
            Log.InfoFormat("Starting receive for client: {0}", _socket.RemoteEndPoint);
            _socket.BeginReceive(_buffer, 0, BufferSize, 0, ReadCallback, null);
        }

        private void ReadCallback(IAsyncResult ar)
        {
            try
            {
                // Read data from the client socket.   
                int bytesRead = _socket.EndReceive(ar);

                if (bytesRead <= 0)
                {
                    Log.InfoFormat("Got 0 byte message. Closing connection: {0}", _socket.RemoteEndPoint);
                    _socket.Close();
                    return;
                }

                TryParseRemaining(_buffer, 0, bytesRead);

                _socket.BeginReceive(_buffer, 0, BufferSize, 0, ReadCallback, null);
            }
            catch (Exception e)
            {
                Log.ErrorFormat("Receive error: {0}", e.Message);
            }
        }

        private void TryParseRemaining(byte[] arr, int start, int length)
        {
            for (int i = start; i < length; i++)
            {
                if (arr[i] == '\n')
                {
                    int len = i + 1 - start;
                    _sb.Append(Encoding.ASCII.GetString(arr, start, len));
                    string res = _sb.ToString();
                    _sb.Clear();

                    TryParseRemaining(arr, i + 1, length);
                    ParseAndRun(res);
                    return;
                }
            }

            // Queue remainder and stop for next time
            if (length - start > 0)
                _sb.Append(Encoding.ASCII.GetString(arr, start, length - start));
        }

        private void ParseAndRun(string str)
        {
            try
            {
                ClientAction action = JsonConvert.DeserializeObject<ClientAction>(str.Trim('\n'));
                if (action == null || action.Type == ClientAction.ActionType.Unknown)
                {
                    ReplyPing();
                }
                else if (action.Type == ClientAction.ActionType.Query)
                {
                    Log.InfoFormat("Received state query");
                    SendState(_manager.GetStateForTimelineSlot(action.TimelineSlot));
                }
                else
                {
                    Log.InfoFormat("Received action: {0}", action);
                    RunAction(action);
                }
            }
            catch (Exception e)
            {
                Log.ErrorFormat("Action run error: {0}", e.Message);
            }
        }

        private void RunAction(ClientAction action)
        {
            switch (action.Type)
            {
                case ClientAction.ActionType.Kill:
                    _manager.KillTimeline(action.TimelineSlot);
                    break;

                case ClientAction.ActionType.Load:
                    if (_manager.LoadTimeline(action.TimelineSlot, action.TimelineFile, action.InstanceName ?? ""))
                        _manager.StartTimeline(action.TimelineSlot, action.Parameters.ToImmutableDictionary());
                    break;

                case ClientAction.ActionType.Cue:
                    _manager.TriggerCue(action.TimelineSlot);
                    break;

                case ClientAction.ActionType.RunChild:
                    _manager.TriggerChild(action.TimelineSlot, action.InstanceName, action.Parameters);
                    break;

                case ClientAction.ActionType.Query:
                    // Nothing to do
                    break;

                default:
                    Log.ErrorFormat("Unknown action type: {0}", action.Type);
                    break;
            }
        }

        public void Close()
        {
            lock (_sendLock)
            {
                _connected = false;
                try
                {
                    _socket?.Close();
                    Log.InfoFormat("Closing client connection");
                }
                catch (IOException)
                {
                }
                _socket = null;
            }
        }

        internal bool SendState(TimelineState state)
        {
            if (state == null)
                return true;

            if (!IsConnected)
                return false;


            Task.Run(() => SendObject(state));
            return true;
        }

        private void SendObject(object obj)
        {
            lock (_sendLock)
            {
                try
                {
                    _socket.Send(Encoding.ASCII.GetBytes(JsonConvert.SerializeObject(obj)));
                }
                catch (Exception)
                {
                    try
                    {
                        _socket?.Close();
                    }
                    catch (IOException)
                    {
                    }
                    _socket = null;
                }
            }
        }

        private bool ReplyPing()
        {
            if (!IsConnected)
                return false;

            Task.Run(() => SendObject(new object()));
            return true;
        }

        public bool IsConnected => _socket != null && _connected && _socket.Connected;

        internal void SendCompleteState()
        {
            _manager.GetCompleteState().ForEach(s => SendState(s));
        }
    }
}